transformer_engine/pytorch/csrc/comm_gemm_overlap.h

```diff
@@ -476,18 +477,19 @@ struct UbufP2PCommOverlap : torch::CustomClassHolder {
         if (i < num_steps - 1) {
           // P2P communication
           userbuffers_send(_ub_reg, send_offset, _ub_reg, send_offset, comm_bytes * 2, _ub_comm,
-                           next_rank, (cudaStream_t)_stream_comm);
+                           next_rank, (cudaStream_t)_stream_send);
           userbuffers_recv(_ub_reg, recv_offset, _ub_reg, recv_offset, comm_bytes * 2, _ub_comm,
-                           prev_rank, (cudaStream_t)_stream_comm);
-          CHECK_CUDA(cudaEventRecord(_stop_comm, (cudaStream_t)_stream_comm));
+                           prev_rank, (cudaStream_t)_stream_recv);
+          CHECK_CUDA(cudaEventRecord(_stop_recv, (cudaStream_t)_stream_recv));
+          CHECK_CUDA(cudaStreamWaitEvent((cudaStream_t)_stream_send, _stop_recv, 0));
           CHECK_CUDA(cudaStreamWaitEvent(
-              (cudaStream_t)_stream_compute[(i + 1) % _stream_compute.size()], _stop_comm, 0));
+              (cudaStream_t)_stream_compute[(i + 1) % _stream_compute.size()], _stop_recv, 0));
         } else if (B_copy.numel() > 0) {
           assert(B_copy.numel() == _ubufs[_tp_id].numel());
           assert(B_copy.element_size() == _ubufs[_tp_id].element_size());
           CHECK_CUDA(cudaMemcpyAsync(B_copy.data_ptr(), _ubufs[_tp_id].data_ptr(),
                                      _ubufs[_tp_id].numel() * _ubufs[_tp_id].element_size(),
-                                     cudaMemcpyDeviceToDevice, (cudaStream_t)_stream_comm));
+                                     cudaMemcpyDeviceToDevice, (cudaStream_t)_stream_send));
```

Comment:
Should we synchronize the main stream on `_stream_send` before `split_overlap_ag` exits to ensure the values in `B_copy` are correct? Alternatively, should this memcpy be on a compute stream?